### ISSUE TITLE
How can I verify that symweb publishing of windows pdbs is happening

### ISSUE THREAD

**Issue by rynowak** [dotnet/arcade#1499]
Hi Folks,

I'm trying to track down whether/if my windows symbols are being published to symweb (for a VS insertion). 

I'm currently looking through logs, and it looks like [this line](https://github.com/dotnet/arcade/blob/531cc5d5d68c0d8b7a6715d85ed085b10c38cac5/src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj#L26) is setting `$(PublishToSymbolServer)` to false because `$(AzureFeedUrl)` has a value. 

Then further down in the Publish target, the `PublishSymbols` task is skipped because `$(PublishToSymbolServer)` is false. 

Is this expected? Is this right? 

If this is not a bug, could you explain or document how this logic is supposed to work?

**Comment by tmat** [comment]
+ @mmitche 

There are two kind of builds - one executed by DNCENG that's publishing all assets to Azure blob storage, and the other that's publishing assets directly. The latter is used for repos that are part of .NET Core stack and allows us control publishing of the entire stack. The other is for repos that build independently of .NET Core stack (e.g. dotnet/interactive-window).

**Comment by tmat** [comment]
BTW: To check that Windows PDBs were published and are available:
`\\symbols\tools\x86\symchk.exe /s SRV*http://symweb "[PDB PATH]"` for checking internally.
`\\symbols\tools\x86\symchk.exe /s SRV*http://msdl.microsoft.com/download/symbols "[PDB PATH]"` for checking externally.

**Comment by rynowak** [comment]
> The former is used for repos that are part of .NET Core stack and allows us control publishing of the entire stack. The latter is for repos that build independently of .NET Core stack (e.g. dotnet/interactive-window)

aspnet/AspNetCore-Tooling is both of these. We build components in this repo for both .NET Core and VS. Is there a reason why we *can't* publish to both of these from the same build?

**Comment by tmat** [comment]
You can have an alternative build leg that switches to the other one (does not pass `DotNetPublishToBlobFeed`).

The reason is that builds that build .NET Core stack need to publish all artifacts at once. We can't have some repos publishing some artifacts and other repos not publishing them. Needs to be all or none.
@mmitche can provide more details.

If we want the same build that's produced by .NET Core stack build to be inserted into VS then we can do that as part of a release pipeline. That would publish symbols from Azure blob storage and then insert into VS.

**Comment by rynowak** [comment]
Thanks Tomas, that's helpful.

So I don't have a requirement that insert the *same build* into both VS and .NET Core. Would the fastest way to get this unblocked to define an additional build pipeline for the repo?

**Comment by tmat** [comment]
I think so. Just set `DotNetPublishToBlobFeed` to false and it should be publishing directly.

**Comment by mmitche** [comment]
@tmat Could @rynowak just run two sets of steps in the same job that pass different parameters?

On the general topic of publishing from the build:

I think these all need to be release pipelines, and we are currently working towards moving these kind of steps out of the build and into release pipelines, starting with packages. Publishing shouldn't be done from the build long term.  It works today for 3.0 because we're not in servicing mode, and we have no security fixes to make.  As soon as we do, we run into a problem where you have to be extremely careful about how builds are launched to avoid publishing artifacts you don't want (stabilized assets, internal-only fixes, etc.).  This is very prone to error.

Instead, we'll change publishing to simply organize build outputs as Azure DevOps artifacts, then launch release pipelines (mostly automatically) based on the intention of build (internal servicing vs. Net Core 3 Daily Public Dev, etc.).  These release pipelines are then tasked with putting the build outputs in their final resting place as well as other tasks (e.g. VS insertion).

See: https://github.com/dotnet/arcade/blob/master/Documentation/CorePackages/AsyncPublishing.md

**Comment by tmat** [comment]
I agree we need to move publishing to release pipelines for .NET Core stack build. 

I would not do that for other builds (repos). At least not unless the release definition is stored in a file that can be shared across all of them. Otherwise it would be nightmare to maintain and too complicated for simple repos that have never needed to deal with any of the problems you mentioned.

**Comment by tmat** [comment]
> Could @rynowak just run two sets of steps in the same job that pass different parameters?

I don't think that's gonna work well. We shouldn't be publishing the same PDB file twice - thru Azure blob storage and directly to the symbol server.

**Comment by mmitche** [comment]
@tmat We think we can just take a single (or few) release def and use it across multiple builds. They're not defined in source, but as long as we minimize the number floating around and they don't end up custom for each repo, we'll be okay.

**Comment by natemcmaster** [comment]
Chatted with @rynowak in person about this as it's blocking product work. What we're going to do to unblock him is:

* Change the AspNetCore-Tooling build definition to publish assets directly to symbweb. The release-pipelines version of the publish symbols task doesn't work in dnceng.
* Use our AzDO release pipeline to push packages from this feed to myget, and continue to manually insert builds into dotnet/toolset. It's not clear to use yet what publishing to the Azure blob storage provides.

It sounds to me like Arcade support for publishing assets is still in flux, and while we have ideas for how things might work someday, we need to get stuff working now with what we have.

cc @markwilkie @Eilon

**Comment by markwilkie** [comment]
@rynowak - Any chance you could share what you ended up implementing?

**Comment by rynowak** [comment]
Yeah, this is the fix: https://github.com/aspnet/AspNetCore-Tooling/commit/5554f10d5be2eadf2c0cfcb299df0b6824ba7161

https://github.com/aspnet/AspNetCore-Tooling/blob/master/azure-pipelines.yml - most of what's here is straight from the examples.

There isn't really an implementation to share 😆 - dotnet/Arcade already knows how to publish symbols to BAR or to symweb (but not both at the same time). I don't need BAR yet, so I disabled that and now it goes to symweb.

**Comment by rynowak** [comment]
@tmat @mmitche - I'm trying to add BAR publishing to this and ending up running into issues that I don't understand. https://dnceng.visualstudio.com/internal/_build/results?buildId=60299

Running symbol publishing as a separate step just doesn't seem to work.

**Comment by markwilkie** [comment]
@rynowak - is this still an issue?

**Comment by rynowak** [comment]
I have a workaround that unblocks me.

I still find it really vexing that the blessed configuration supports publishing to BAR **OR** Symweb but not both. I guess it's up to you and @mmitche if you want to track that as an improvement. I know this is all planned to change eventually.

**Comment by mmitche** [comment]
@tmat may have a better idea of why this decision was made.

**Comment by mmitche** [comment]
What's your workaround, btw? IIRC you can set a property to force publishing symbols.

**Comment by rynowak** [comment]
My workaround is to set both properties to true in the yaml file. https://github.com/aspnet/AspNetCore-Tooling/blob/master/azure-pipelines.yml#L64

This isn't the documented/supported way to configure these things though, it's just something that happens to work. See my high-level feedback here:  https://github.com/dotnet/arcade/issues/1616

**Comment by markwilkie** [comment]
I'm adding this to the publishing epic to be sure we get the right resolution on this.

**Comment by JohnTortugo** [comment]
Closing this issue as its a duplicate of #173 and #1874
